File: pe-core/src/main/java/org/saar/core/common/inference/weak/WeakVertex.java
package org.saar.core.common.inference.weak;

import org.saar.core.mesh.Vertex;
import org.saar.lwjgl.opengl.objects.attributes.IAttribute;
import org.saar.lwjgl.opengl.primitive.GlPrimitive;
import org.saar.lwjgl.util.DataWriter;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class WeakVertex implements Vertex {

    private final List<GlPrimitive> primitives = new ArrayList<>();

    public WeakVertex with(GlPrimitive primitive) {
        this.primitives.add(primitive);
        return this;
    }

    public void write(DataWriter writer) {
        for (GlPrimitive primitive : this.primitives) {
            primitive.write(writer);
        }
    }

    public IAttribute[] getAttributes() {
        int index = 0;
        final List<IAttribute> attributes = new ArrayList<>();
        for (GlPrimitive primitive : this.primitives) {
            Collections.addAll(attributes, primitive
                    .attribute(index++, false, 0));
        }
        return attributes.toArray(new IAttribute[0]);
    }
}

File: pe-core/src/main/java/org/saar/core/common/inference/weak/WeakVertex.kt
package org.saar.core.common.inference.weak

import org.saar.core.mesh.Vertex
import org.saar.lwjgl.opengl.primitive.GlPrimitive
import org.saar.lwjgl.util.DataWriter

class WeakVertex(private val primitives: Iterable<GlPrimitive>) : Vertex {

    fun write(writer: DataWriter) = this.primitives.forEach { it.write(writer) }

    val attributes = this.primitives.flatMapIndexed { index, primitive ->
        primitive.attribute(index, false, 0).asIterable()
    }

    class Builder {
        private val primitives = mutableListOf<GlPrimitive>()

        fun with(primitive: GlPrimitive): Builder {
            this.primitives.add(primitive)
            return this
        }

        fun build() = WeakVertex(this.primitives)
    }
}
